grep-0021-pmt-replacement.md

```diff
@@ -0,0 +1,90 @@
+# GREP 0019 -- Restructure GNU Radio Blocks
+
+- Original Author: Josh Morman <mormjb@gmail.com>
+- Champion: Josh Morman <mormjb@gmail.com>
+- Status: Draft
+
+History:
+
+- 25-Feb-2021: Initial Draft
+
+## Abstract
+
+This GREP proposes replacing the PMT types with classes that support better 
+serialization and have a cleaner API
+
+## Copyright / License
+
+CC-BY-ND
+
+## Motivation
+
+The existing PMT API is confusing, doesn't take advantage of OOP concepts, and
+the nomenclature for doing simple things can be inconsistent or lacking.  Additionally,
+there has been work to show that the serialization/deserialization is inherently slow
+and specific to the PMT library.
+
+The leading candidate to natively support serialization of polymorphic types is
+[flatbuffers](https://google.github.io/flatbuffers/).  This library provides cross
+platform serialization and deserialization and is efficient in storing the serialized
+representation in a flat buffer that can be accessed directly
+
+## Description
+
+Create a new in-tree module - `pmtf` - that will live alongside `pmt` during a 
```

Comment:
Will GR PMTFs will be able to be understood by outside software without the need for libpmtf? If so, and i think this would be an important objective, we should state it here. Alternatively, if libpmtf is going to be necessary it might be worth starting it as an external library from the beginning to make GR more readily integrate into larger projects.

Reply:
I guess there's no reason why they wouldn't be understood outside, though libpmtf would make it much easier to parse.  Using the flatbuffers API directly while possible is fairly cumbersome.  Most of libpmtf prototype is just wrappers around the flatbuffers API